content/blog/purrr-walk-this-way/index.Rmd

```diff
@@ -0,0 +1,229 @@
+* [ ] `usethis::use_tidy_thanks()`
+-->
+
+## Meet the `map()` family
+
+purrr's `map()` family of functions are tools for **iteration**, performing the same action on multiple inputs.
+If you're new to purrr, the [Iteration chapter](https://r4ds.had.co.nz/iteration.html#iteration) of R for Data Science is a good place to get started.
+
+One of the benefits of using `map()` is that the function has variants (e.g. `map2()`, `pmap()`, etc.) all of which work the same way.
+To borrow from Jennifer Thompson's excellent [Intro to purrr](https://github.com/jenniferthompson/RLadiesIntroToPurrr),the arguments can be broken into two groups: what we're iterating over, and what we're doing each time.
+The adapted figure below shows what this looks like for `map()`, `map2()`, and `pmap()`.
+
+```{r}
+#| label: map-args
+#| echo: FALSE
+#| fig.align: 'center'
+#| fig.cap: Grouped map function arguments, adapted from Intro to purrr by Jennifer Thompson
+#| fig.alt: "Highlighted titles read: what we're iterating over, and what we're doing each time. For map(.x = , .f = ) .x is what we're iterating over and .f is what we're doing each time. For map2(.x = , .y = , .f = ) .x and .y are what we're iterating over and .f is what we're doing each time. For pmap(.l = list(), .f = ) .l is what we're iterating over and .f is what we're doing each time."
+knitr::include_graphics("purrr-map-args.png")
+```
+
+In addition to handling different input arguments, the map family of functions has variants that create different outputs.
+The following table from the [Map-variants section of Advanced R](https://adv-r.hadley.nz/functionals.html#map-variants) shows how the orthogonal inputs and outputs can be used to organise the variants into a matrix:
+
+|                       | List     | Atomic            | Same type   | Nothing   |
+|-----------------------|----------|-------------------|-------------|-----------|
+| One argument          | `map()`  | `map_lgl()`, ...  | `modify()`  | `walk()`  |
+| Two arguments         | `map2()` | `map2_lgl()`, ... | `modify2()` | `walk2()` |
+| One argument + index  | `imap()` | `imap_lgl()`, ... | `imodify()` | `iwalk()` |
+| N arguments           | `pmap()` | `pmap_lgl()`, ... | ---         | `pwalk()` |
+
+
+## What's up with `walk()`?
+
+Based on the table above, you might think that `walk()` isn't very useful.
+Indeed, `walk()`, `walk2()`, and `pwalk()` all invisibly return `.x`.
+However, they come in handy when you want to call a function for its ***side effects*** rather than its return value.
+
+Here, we'll go through two common use cases: saving multiple CSVs, and multiple plots.
+We'll also make use of the [fs](https://fs.r-lib.org/) package, a cross-platform interface to file system operations, to inspect our outputs.
+
+If you want to try this out but don't want to save files locally, there's a [companion project on **Posit Cloud**](https://posit.cloud/content/5983147) where you can follow along.
```

Comment:
Would it be better to use `.qmd` files on Cloud? Can we make the project open with one of them already open?

Reply:
Done.